File: day03/main.py
from string import ascii_lowercase, ascii_uppercase

priorities = ascii_lowercase + ascii_uppercase

filename = "./input"

with open(filename, "r") as file:
    lines = file.readlines()
rucksacks = [line.strip() for line in lines]


priorities_sum = 0
for rucksack in rucksacks:
    middle = len(rucksack) // 2
    c1 = rucksack[:middle]
    c2 = rucksack[middle:]

    inter = set(c1).intersection(c2)

    priority = priorities.index(inter.pop()) + 1
    priorities_sum += priority

print(f"Puzzle 1: {priorities_sum}")


priorities_sum = 0
for i in range(0, len(rucksacks), 3):
    group_rucksacks = rucksacks[i : i + 3]

    badge = (
        set(group_rucksacks[0])
        .intersection(group_rucksacks[1])
        .intersection(group_rucksacks[2])
    )
    priority = priorities.index(badge.pop()) + 1
    priorities_sum += priority

print(f"Puzzle 2: {priorities_sum}")

File: day03/main.go
package main

import (
	aoc "aoc/utils"
	"errors"
	"fmt"
)

const PRIORITIES = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"

func findDuplicate(c1, c2 string) (rune, error) {
	for _, item1 := range c1 {
		for _, item2 := range c2 {
			if (item1 == item2) {
				return item1, nil
			}
		}
	}
	return 0, errors.New("No duplicate item found")
}

func getPriority(item rune) (int, error) {
	for i, ch := range PRIORITIES {
		if (ch == item) {
			return i + 1, nil
		}
	}
	return 0, errors.New("Unable to prioritize item")
}

func main() {
	rucksacks, err := aoc.ReadLines("./input")
	aoc.Handle(err)

	var prioritiesSum int
	for _, rucksack := range rucksacks {
		middle := len(rucksack) / 2
		c1 := rucksack[:middle]
		c2 := rucksack[middle:]

		duplicate, _ := findDuplicate(c1, c2)
		
		priority, _ := getPriority(duplicate)
		prioritiesSum += priority
	}
	fmt.Println("Puzzle 1:", prioritiesSum)

	prioritiesSum = 0
	for i := 0; i < len(rucksacks); i += 3 {
		rucksackGroup := rucksacks[i:i+3]

		var badge rune
		for _, item1 := range rucksackGroup[0] {
			for _, item2 := range rucksackGroup[1] {
				for _, item3 := range rucksackGroup[2] {
					if (item1 == item2 && item1 == item3) {
						badge = item1
					}
				}
				
			}
		}
		priority, _ := getPriority(badge)
		prioritiesSum += priority
	}
	fmt.Println("Puzzle 2:", prioritiesSum)
}

